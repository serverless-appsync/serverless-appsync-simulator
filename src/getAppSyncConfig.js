import { AmplifyAppSyncSimulatorAuthenticationType as AuthTypes } from '@aws-amplify/amplify-appsync-simulator';
import axios from 'axios';
import fs from 'fs';
import { forEach, isNil, first } from 'lodash';
import path from 'path';
import { mergeTypeDefs } from '@graphql-tools/merge';
import * as globby from 'globby';
import directLambdaRequest from './templates/direct-lambda.request.vtl';
import directLambdaResponse from './templates/direct-lambda.response.vtl';
import {
  DEFAULT_MAPPING_TEMPLATE_LOCATION,
  DEFAULT_ENCODING,
  DEFAULT_SCHEMA_FILE,
  DEFAULT_HTTP_METHOD,
  DEFAULT_RESOLVER_TYPE,
  HTTPMessage,
  MappingTemplateType,
  SourceType,
} from './constants';

const directLambdaMappingTemplates = {
  request: directLambdaRequest,
  response: directLambdaResponse,
};

const convertAppSyncConfigBackToOldVersion = (appSyncConfig) => {
  if (
    appSyncConfig.authentication &&
    appSyncConfig.authentication.hasOwnProperty('type')
  ) {
    appSyncConfig.authenticationType = appSyncConfig.authentication.type;
  }

  if (appSyncConfig.dataSources && !Array.isArray(appSyncConfig.dataSources)) {
    console.log('AppSync Simulator: Convert new configuration to old one');
    const arrayDataSources = [];
    for (const [key, value] of Object.entries(appSyncConfig.dataSources)) {
      const name = value.hasOwnProperty('name') ? value.name : key;

      arrayDataSources.push({
        ...value,
        name: name,
      });
    }
    appSyncConfig.dataSources = arrayDataSources;
    appSyncConfig.version = '2.2.0';
  }

  if (
    appSyncConfig.pipelineFunctions &&
    !Array.isArray(appSyncConfig.pipelineFunctions)
  ) {
    console.log('AppSync Simulator: Convert new configuration to old one');
    const arrayPipelineFunctions = [];
    for (const [key, value] of Object.entries(
      appSyncConfig.pipelineFunctions,
    )) {
      const name = value.hasOwnProperty('name') ? value.name : key;
      const pipelineFunction = {
        ...value,
        name: name,
      };

      if (
        !pipelineFunction.hasOwnProperty('code') &&
        !(
          pipelineFunction.hasOwnProperty('request') ||
          pipelineFunction.hasOwnProperty('response')
        )
      ) {
        if (!pipelineFunction.hasOwnProperty('request')) {
          pipelineFunction['request'] = false;
        }
        if (!pipelineFunction.hasOwnProperty('response')) {
          pipelineFunction['response'] = false;
        }
      }

      arrayPipelineFunctions.push(pipelineFunction);
    }
    appSyncConfig.functionConfigurations = arrayPipelineFunctions;
    appSyncConfig.version = '2.2.0';
  }

  if (appSyncConfig.resolvers) {
    console.log('AppSync Simulator: Convert new configuration to old one');
    const arrayMappingTemplates = [];

    for (const [key, value] of Object.entries(appSyncConfig.resolvers)) {
      const keySplitted = key.split('.');
      const type = value.hasOwnProperty('type') ? value.type : keySplitted[0];
      const field = value.hasOwnProperty('field')
        ? value.field
        : keySplitted[1];

      arrayMappingTemplates.push({
        ...value,
        type: type,
        field: field,
      });
    }
    // console.log('Mapping Config', arrayMappingTemplates);
    appSyncConfig.mappingTemplates = arrayMappingTemplates;
    appSyncConfig.version = '2.2.0';
  }

  return appSyncConfig;
};

export default function getAppSyncConfig(context, appSyncConfig) {
  // Flattening params

  const modifiedAppSyncConfig =
    convertAppSyncConfigBackToOldVersion(appSyncConfig);

  const cfg = {
    ...modifiedAppSyncConfig,
    mappingTemplates: (modifiedAppSyncConfig.mappingTemplates || []).flat(),
    functionConfigurations: (
      modifiedAppSyncConfig.functionConfigurations || []
    ).flat(),
    dataSources: (modifiedAppSyncConfig.dataSources || []).flat(),
  };

  const mappingTemplatesLocation = path.join(
    context.serverless.config.servicePath,
    modifiedAppSyncConfig.hasOwnProperty('version')
      ? ''
      : cfg.mappingTemplatesLocation || DEFAULT_MAPPING_TEMPLATE_LOCATION,
  );

  const functionConfigurationsLocation = path.join(
    context.serverless.config.servicePath,
    modifiedAppSyncConfig.hasOwnProperty('version')
      ? ''
      : cfg.functionConfigurationsLocation ||
          cfg.mappingTemplatesLocation ||
          DEFAULT_MAPPING_TEMPLATE_LOCATION,
  );

  const { defaultMappingTemplates = {} } = cfg;

  const getMappingTemplate = (filePath, type) => {
    switch (type) {
      case MappingTemplateType.MAPPING_TEMPLATE:
        return fs.readFileSync(path.join(mappingTemplatesLocation, filePath), {
          encoding: DEFAULT_ENCODING,
        });
      case MappingTemplateType.FUNCTION_CONFIGURATION:
        return fs.readFileSync(
          path.join(functionConfigurationsLocation, filePath),
          {
            encoding: DEFAULT_ENCODING,
          },
        );
      default:
        return null;
    }
  };

  const toAbsolutePosixPath = (basePath, filePath) =>
    (path.isAbsolute(filePath)
      ? filePath
      : path.join(basePath, filePath)
    ).replace(/\\/g, '/');

  const globFilePaths = (basePath, filePaths) => {
    return filePaths
      .map((filePath) => {
        const paths = globby.sync(toAbsolutePosixPath(basePath, filePath));
        if (path.isAbsolute(filePath)) {
          return paths;
        } else {
          // For backward compatibility with FileMap, revert to relative path
          return paths.map((p) => path.relative(basePath, p));
        }
      })
      .flat();
  };

  const getFileMap = (basePath, filePath) => ({
    path: filePath,
    content: fs.readFileSync(toAbsolutePosixPath(basePath, filePath), {
      encoding: DEFAULT_ENCODING,
    }),
  });

  const makeDataSource = (source) => {
    if (source.name === undefined || source.type === undefined) {
      return null;
    }

    const dataSource = {
      name: source.name,
      type: source.type,
    };

    switch (source.type) {
      case SourceType.AMAZON_DYNAMODB: {
        return {
          ...dataSource,
          config: {
            ...context.options.dynamoDb,
            tableName: source.config.tableName,
          },
        };
      }
      case SourceType.RELATIONAL_DATABASE: {
        return {
          ...dataSource,
          rds: context.options.rds,
        };
      }
      case SourceType.AWS_LAMBDA: {
        const { functionName } = source.config;
        if (functionName === undefined) {
          context.plugin.log(`${source.name} does not have a functionName`, {
            color: 'orange',
          });
          return null;
        }

        const conf = context.options;
        const func =
          conf.functions?.[functionName] ||
          context.serverless.service.functions?.[functionName];

        if (func === undefined) {
          context.plugin.log(`The ${functionName} function is not defined`, {
            color: 'orange',
          });
          return null;
        }

        let url, method;
        if (func.url) {
          url = func.url;
          method = func.method;
        } else {
          url = `${context.options.httpProtocol}://${context.options.httpHost}:${context.options.lambdaPort}/2015-03-31/functions/${func.name}/invocations`;
        }
        return {
          ...dataSource,
          invoke: async (payload) => {
            const result = await axios.request({
              url,
              method: method || DEFAULT_HTTP_METHOD,
              data: payload,
              headers: payload?.request?.headers,
              validateStatus: false,
            });
            // When the Lambda returns an error, the status code is 200 OK.
            // The presence of an error is indicated by a header in the response.
            // 400 and 500-series status codes are reserved for invocation errors:
            // https://docs.aws.amazon.com/lambda/latest/dg/API_Invoke.html#API_Invoke_Errors
            if (result.status === 200) {
              const errorType =
                result.headers['x-amz-function-error'] ||
                result.headers['X-Amz-Function-Error'] ||
                result.headers['x-amzn-errortype'] ||
                result.headers['x-amzn-ErrorType'];
              if (errorType) {
                throw {
                  type: `Lambda:${errorType}`,
                  message: result.data.errorMessage,
                };
              }
              // If the result of a lambda function is null or undefined, it returns as an empty string via HTTP.
              // Then, AppSync handles an empty string of the response as null (or undefined).
              if (result.data === '') {
                return null;
              }
              return result.data;
            } else {
              throw new Error(
                `Request failed with status code ${result.status}`,
              );
            }
          },
        };
      }
      case SourceType.AMAZON_ELASTICSEARCH:
        return {
          ...context.options.openSearch,
          ...dataSource,
          endpoint: source.config.endpoint,
        };
      case SourceType.HTTP: {
        return {
          ...dataSource,
          endpoint: source.config.endpoint,
        };
      }
      default:
        return dataSource;
    }
  };

  const makeMappingTemplate = (resolver, type, templateType) => {
    const { name, type: parent, field, substitutions = {} } = resolver;

    const defaultTemplatePrefix = name || `${parent}.${field}`;
    const templatePath = !isNil(resolver?.[type])
      ? resolver?.[type]
      : !isNil(defaultMappingTemplates?.[type])
      ? defaultMappingTemplates?.[type]
      : `${defaultTemplatePrefix}.${type}.vtl`;

    let mappingTemplate;
    // Direct lambda
    // For direct lambdas, we use a default mapping template
    // See https://amzn.to/3ncV3Dz
    if (templatePath === false) {
      mappingTemplate = directLambdaMappingTemplates[type];
    } else {
      mappingTemplate = getMappingTemplate(templatePath, templateType);
      // Substitutions
      const allSubstitutions = { ...cfg.substitutions, ...substitutions };
      forEach(allSubstitutions, (value, variable) => {
        const regExp = new RegExp(`\\$\{?${variable}}?`, 'g');
        mappingTemplate = mappingTemplate.replace(regExp, value);
      });
    }

    return mappingTemplate;
  };

  const makeResolver = (resolver) => {
    let templateType = MappingTemplateType.MAPPING_TEMPLATE;
    return {
      kind: resolver.kind || DEFAULT_RESOLVER_TYPE,
      fieldName: resolver.field,
      typeName: resolver.type,
      dataSourceName: resolver.dataSource,
      functions: resolver.functions,
      requestMappingTemplate: makeMappingTemplate(
        resolver,
        HTTPMessage.REQUEST,
        templateType,
      ),
      responseMappingTemplate: makeMappingTemplate(
        resolver,
        HTTPMessage.RESPONSE,
        templateType,
      ),
    };
  };

  const makeFunctionConfiguration = (config) => {
    let templateType = MappingTemplateType.FUNCTION_CONFIGURATION;
    return {
      dataSourceName: config.dataSource,
      name: config.name,
      requestMappingTemplate: makeMappingTemplate(
        config,
        HTTPMessage.REQUEST,
        templateType,
      ),
      responseMappingTemplate: makeMappingTemplate(
        config,
        HTTPMessage.RESPONSE,
        templateType,
      ),
    };
  };

  const makeAuthType = (authType) => {
    const auth = {
      authenticationType: authType.authenticationType,
    };

    if (auth.authenticationType === AuthTypes.AMAZON_COGNITO_USER_POOLS) {
      auth.cognitoUserPoolConfig = {
        AppIdClientRegex: authType.userPoolConfig.appIdClientRegex,
      };
    } else if (auth.authenticationType === AuthTypes.OPENID_CONNECT) {
      auth.openIDConnectConfig = {
        Issuer: authType.openIdConnectConfig.issuer,
        ClientId: authType.openIdConnectConfig.clientId,
      };
    }

    return auth;
  };

  const makeAppSync = (config) => ({
    name: config.name,
    apiKey: context.options.apiKey,
    defaultAuthenticationType: makeAuthType(config),
    additionalAuthenticationProviders: (
      config.additionalAuthenticationProviders || []
    ).map(makeAuthType),
  });

  // Load the schema. If multiple provided, merge them
  const schemaPaths = Array.isArray(cfg.schema)
    ? cfg.schema
    : [cfg.schema || DEFAULT_SCHEMA_FILE];
  const basePath = context.serverless.config.servicePath;
  const schemas = globFilePaths(basePath, schemaPaths).map((schemaPath) =>
    getFileMap(basePath, schemaPath),
  );
  const schema = {
    path: first(schemas).path,
    content: mergeTypeDefs(
      schemas.map((s) => s.content),
      {
        useSchemaDefinition: true,
        forceSchemaDefinition: true,
        throwOnConflict: true,
        commentDescriptions: true,
        reverseDirectives: true,
      },
    ),
  };

  return {
    appSync: makeAppSync(cfg),
    schema,
    resolvers: cfg.mappingTemplates.map(makeResolver),
    dataSources: cfg.dataSources.map(makeDataSource).filter((v) => v !== null),
    functions: cfg.functionConfigurations.map(makeFunctionConfiguration),
  };
}
