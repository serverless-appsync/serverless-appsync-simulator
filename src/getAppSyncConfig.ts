import { forEach, merge } from 'lodash';
import { AppSyncConfigInput } from './types/appSync';
import {
  ApiKeyConfig,
  AppSyncConfig,
  DataSourceConfig,
  PipelineFunctionConfig,
  ResolverConfig,
} from './types/plugin';


const flattenMaps = <T>(
  input?: Record<string, T> | Record<string, T>[],
): Record<string, T> => {
  if (Array.isArray(input)) {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
    return merge({}, ...input);
  } else {
    return merge({}, input);
  }
};

const isUnitResolver = (resolver: {
  kind?: 'UNIT' | 'PIPELINE';
}): resolver is { kind: 'UNIT' } => {
  return resolver.kind === 'UNIT';
};

const isPipelineResolver = (resolver: {
  kind?: 'UNIT' | 'PIPELINE';
}): resolver is { kind: 'PIPELINE' } => {
  return !resolver.kind || resolver.kind === 'PIPELINE';
};

const toResourceName = (name: string) => {
  return name.replace(/[^a-z_]/i, '_');
};

export const getAppSyncConfig = (config: AppSyncConfigInput): AppSyncConfig => {
  const schema = Array.isArray(config.schema)
    ? config.schema
    : [config.schema || 'schema.graphql'];

  const dataSources: Record<string, DataSourceConfig> = {};
  const resolvers: Record<string, ResolverConfig> = {};
  const pipelineFunctions: Record<string, PipelineFunctionConfig> = {};

  forEach(flattenMaps(config.dataSources), (ds, name) => {
    dataSources[name] = {
      ...ds,
      name,
    };
  });

  forEach(flattenMaps(config.resolvers), (resolver, typeAndField) => {
    const [type, field] = typeAndField.split('.');
    if (type == null || field == null) {
      throw new Error(
        `Invalid resolver key. Must be the format([Type].[Field]): ${typeAndField}`,
      );
    }

    if (typeof resolver === 'string') {
      resolvers[typeAndField] = {
        dataSource: resolver,
        kind: 'UNIT',
        type,
        field,
      };
      return;
    }

    if (isUnitResolver(resolver) && typeof resolver.dataSource === 'object') {
      const name = typeAndField.replace(/[^a-z_]/i, '_');
      dataSources[name] = {
        ...resolver.dataSource,
        name,
      };
    }

    resolvers[typeAndField] = {
      ...resolver,
      type: resolver.type || type,
      field: resolver.field || field,
      ...(isUnitResolver(resolver)
        ? {
            kind: 'UNIT',
            dataSource:
              typeof resolver.dataSource === 'object'
                ? typeAndField.replace(/[^a-z_]/i, '_')
                : resolver.dataSource,
          }
        : {
            kind: 'PIPELINE',
            functions: resolver.functions.map((f, index) => {
              if (typeof f === 'string') {
                return f;
              }
              const name = `${toResourceName(typeAndField)}_${index}`;
              pipelineFunctions[name] = {
                ...f,
                name,
                dataSource:
                  typeof f.dataSource === 'string' ? f.dataSource : name,
              };
              if (typeof f.dataSource === 'object') {
                dataSources[name] = {
                  ...f.dataSource,
                  name,
                };
              }
              return name;
            }),
          }),
    };
  });

  forEach(flattenMaps(config.pipelineFunctions), (func, name) => {
    if (typeof func.dataSource === 'object') {
      dataSources[name] = {
        ...func.dataSource,
        name,
      };
    }

    pipelineFunctions[name] = {
      ...func,
      dataSource: typeof func.dataSource === 'string' ? func.dataSource : name,
      name,
    };
  });

  const additionalAuthentications = config.additionalAuthentications || [];

  let apiKeys: Record<string, ApiKeyConfig> | undefined;
  if (
    config.authentication.type === 'API_KEY' ||
    additionalAuthentications.some((auth) => auth.type === 'API_KEY')
  ) {
    const inputKeys = config.apiKeys || [];

    apiKeys = inputKeys.reduce<Record<string, ApiKeyConfig>>((acc, key) => {
      if (typeof key === 'string') {
        acc[key] = { name: key };
      } else {
        acc[key.name] = key;
      }
      return acc;
    }, {});
  }

  return {
    ...config,
    additionalAuthentications,
    apiKeys,
    schema,
    dataSources,
    resolvers,
    pipelineFunctions,
  };
};
