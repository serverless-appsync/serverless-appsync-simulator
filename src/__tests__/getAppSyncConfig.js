import path from 'path';
import getAppSyncConfig from '../getAppSyncConfig';

describe('getAppSyncConfig', () => {
  it('should generate a valid config even with serverless-appsync-plugin >= 2.2.0', () => {
    const config = {
      dataSources: {
        lambda: {
          type: 'AWS_LAMBDA',
          name: 'lambda',
          config: {
            functionName: 'getPosts',
          },
        },
        dynamodb: {
          type: 'AMAZON_DYNAMODB',
          name: 'dynamodb',
          config: {
            tableName: 'myTable',
          },
        },
        http: {
          type: 'HTTP',
          name: 'http',
          config: {
            endpoint: 'http://127.0.0.1',
          },
        },
      },
    };

    const result = getAppSyncConfig(
      {
        options: {
          apiKey: '123456789',
          dynamoDb: {
            endpoint: `http://localhost:8000`,
            region: 'localhost',
            accessKeyId: 'DEFAULT_ACCESS_KEY',
            secretAccessKey: 'DEFAULT_SECRET',
            sessionToken: 'DEFAULT_SESSION_TOKEN',
          },
        },
        serverless: {
          config: { servicePath: path.join(__dirname, 'files') },
          service: {
            functions: {
              getPost: {
                hndler: 'index.handler',
              },
              getPosts: {
                hndler: 'index.handler',
              },
            },
          },
        },
      },
      config,
    );
    expect(result.appSync).toMatchSnapshot();
    expect(result.schema).toMatchSnapshot();
    expect(result.resolvers).toMatchSnapshot();
    expect(result.dataSources).toMatchSnapshot();
    expect(result.functions).toMatchSnapshot();
  });

  it('should generate a valid config', () => {
    const config = {
      name: 'myAPI',
      authenticationType: 'API_KEY',
      schema: '*.graphql',
      defaultMappingTemplates: {
        request: 'default.request.vtl',
        response: 'default.response.vtl',
      },
      mappingTemplates: [
        {
          dataSource: 'lambda',
          type: 'Query',
          field: 'templates',
          request: 'lambda.request.vtl',
          response: 'lambda.response.vtl',
          substitutions: {
            mySubVar: 'lambda',
          },
        },
        {
          dataSource: 'lambda',
          type: 'Query',
          field: 'default',
          substitutions: {
            mySubVar: 'default',
          },
        },
        {
          dataSource: 'lambda',
          type: 'Query',
          field: 'directLambda',
          request: false,
          response: false,
        },
        {
          type: 'Query',
          kind: 'PIPELINE',
          field: 'pipeline',
          functions: ['func', 'func-default'],
          substitutions: {
            mySubVar: 'pipeline',
          },
        },
      ],
      functionConfigurations: [
        {
          dataSource: 'lambda',
          name: 'func',
          request: 'lambda.request.vtl',
          response: 'lambda.response.vtl',
          substitutions: {
            mySubVar: 'template-function',
          },
        },
        {
          dataSource: 'lambda',
          name: 'func-default',
          substitutions: {
            mySubVar: 'default-function',
          },
        },
        {
          dataSource: 'lambda',
          name: 'func-direct',
          request: false,
          response: false,
        },
      ],
      dataSources: [
        {
          type: 'AWS_LAMBDA',
          name: 'lambda',
          config: {
            functionName: 'getPosts',
          },
        },
        {
          type: 'AMAZON_DYNAMODB',
          name: 'dynamodb',
          config: {
            tableName: 'myTable',
          },
        },
        {
          type: 'HTTP',
          name: 'http',
          config: {
            endpoint: 'http://127.0.0.1',
          },
        },
      ],
    };

    const result = getAppSyncConfig(
      {
        options: {
          apiKey: '123456789',
          dynamoDb: {
            endpoint: `http://localhost:8000`,
            region: 'localhost',
            accessKeyId: 'DEFAULT_ACCESS_KEY',
            secretAccessKey: 'DEFAULT_SECRET',
            sessionToken: 'DEFAULT_SESSION_TOKEN',
          },
        },
        serverless: {
          config: { servicePath: path.join(__dirname, 'files') },
          service: {
            functions: {
              getPost: {
                hndler: 'index.handler',
              },
              getPosts: {
                hndler: 'index.handler',
              },
            },
          },
        },
      },
      config,
    );
    expect(result.appSync).toMatchSnapshot();
    expect(result.schema).toMatchSnapshot();
    expect(result.resolvers).toMatchSnapshot();
    expect(result.dataSources).toMatchSnapshot();
    expect(result.functions).toMatchSnapshot();
  });
});
