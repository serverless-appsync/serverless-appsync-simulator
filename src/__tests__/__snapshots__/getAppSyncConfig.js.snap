// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAppSyncConfig should generate a valid config 1`] = `
Object {
  "additionalAuthenticationProviders": Array [],
  "apiKey": "123456789",
  "defaultAuthenticationType": Object {
    "authenticationType": "API_KEY",
  },
  "name": "myAPI",
}
`;

exports[`getAppSyncConfig should generate a valid config 2`] = `
Object {
  "content": "type Post {
  userId: Int!
  id: Int!
  title: String!
  body: String!
}

type Query {
  getPost: Post
  getPosts: [Post]!
}

schema {
  query: Query
}",
  "path": "schema.graphql",
}
`;

exports[`getAppSyncConfig should generate a valid config 3`] = `
Array [
  Object {
    "dataSourceName": "lambda",
    "fieldName": "templates",
    "functions": undefined,
    "kind": "UNIT",
    "requestMappingTemplate": "{
    \\"version\\": \\"2018-05-29\\",
    \\"operation\\": \\"Invoke\\",
    \\"payload\\": {
      \\"substitution\\": \\"lambda\\",
      \\"args\\": $utils.toJson($context.arguments)
    }
}
",
    "responseMappingTemplate": "$utils.toJson($context.result.lambda)
",
    "typeName": "Query",
  },
  Object {
    "dataSourceName": "lambda",
    "fieldName": "default",
    "functions": undefined,
    "kind": "UNIT",
    "requestMappingTemplate": "{
    \\"version\\": \\"2018-05-29\\",
    \\"operation\\": \\"Invoke\\",
    \\"payload\\": {
      \\"substitution\\": \\"default\\",
      \\"type\\": \\"default\\"
    }
}
",
    "responseMappingTemplate": "$utils.toJson($context.result.default)
",
    "typeName": "Query",
  },
  Object {
    "dataSourceName": "lambda",
    "fieldName": "directLambda",
    "functions": undefined,
    "kind": "UNIT",
    "requestMappingTemplate": "## Direct lambda request
{
    \\"version\\": \\"2018-05-29\\",
    \\"operation\\": \\"Invoke\\",
    \\"payload\\": $utils.toJson($context)
}
",
    "responseMappingTemplate": "## Direct lambda response
#if($ctx.error)
    $util.error($ctx.error.message, $ctx.error.type, $ctx.result)
#end
$util.toJson($ctx.result)
",
    "typeName": "Query",
  },
  Object {
    "dataSourceName": undefined,
    "fieldName": "pipeline",
    "functions": Array [
      "func",
      "func-default",
    ],
    "kind": "PIPELINE",
    "requestMappingTemplate": "{
    \\"version\\": \\"2018-05-29\\",
    \\"operation\\": \\"Invoke\\",
    \\"payload\\": {
      \\"substitution\\": \\"pipeline\\",
      \\"type\\": \\"default\\"
    }
}
",
    "responseMappingTemplate": "$utils.toJson($context.result.default)
",
    "typeName": "Query",
  },
]
`;

exports[`getAppSyncConfig should generate a valid config 4`] = `
Array [
  Object {
    "invoke": [Function],
    "name": "lambda",
    "type": "AWS_LAMBDA",
  },
  Object {
    "config": Object {
      "accessKeyId": "DEFAULT_ACCESS_KEY",
      "endpoint": "http://localhost:8000",
      "region": "localhost",
      "secretAccessKey": "DEFAULT_SECRET",
      "sessionToken": "DEFAULT_SESSION_TOKEN",
      "tableName": "myTable",
    },
    "name": "dynamodb",
    "type": "AMAZON_DYNAMODB",
  },
  Object {
    "endpoint": "http://127.0.0.1",
    "name": "http",
    "type": "HTTP",
  },
]
`;

exports[`getAppSyncConfig should generate a valid config 5`] = `
Array [
  Object {
    "dataSourceName": "lambda",
    "name": "func",
    "requestMappingTemplate": "{
    \\"version\\": \\"2018-05-29\\",
    \\"operation\\": \\"Invoke\\",
    \\"payload\\": {
      \\"substitution\\": \\"template-function\\",
      \\"args\\": $utils.toJson($context.arguments)
    }
}
",
    "responseMappingTemplate": "$utils.toJson($context.result.lambda)
",
  },
  Object {
    "dataSourceName": "lambda",
    "name": "func-default",
    "requestMappingTemplate": "{
    \\"version\\": \\"2018-05-29\\",
    \\"operation\\": \\"Invoke\\",
    \\"payload\\": {
      \\"substitution\\": \\"default-function\\",
      \\"type\\": \\"default\\"
    }
}
",
    "responseMappingTemplate": "$utils.toJson($context.result.default)
",
  },
  Object {
    "dataSourceName": "lambda",
    "name": "func-direct",
    "requestMappingTemplate": "## Direct lambda request
{
    \\"version\\": \\"2018-05-29\\",
    \\"operation\\": \\"Invoke\\",
    \\"payload\\": $utils.toJson($context)
}
",
    "responseMappingTemplate": "## Direct lambda response
#if($ctx.error)
    $util.error($ctx.error.message, $ctx.error.type, $ctx.result)
#end
$util.toJson($ctx.result)
",
  },
]
`;

exports[`getAppSyncConfig should generate a valid config even with serverless-appsync-plugin >= 2.2.0 1`] = `
Object {
  "additionalAuthenticationProviders": Array [],
  "apiKey": "123456789",
  "defaultAuthenticationType": Object {
    "authenticationType": undefined,
  },
  "name": undefined,
}
`;

exports[`getAppSyncConfig should generate a valid config even with serverless-appsync-plugin >= 2.2.0 2`] = `
Object {
  "content": "type Post {
  userId: Int!
  id: Int!
  title: String!
  body: String!
}

type Query {
  getPost: Post
  getPosts: [Post]!
}

schema {
  query: Query
}",
  "path": "schema.graphql",
}
`;

exports[`getAppSyncConfig should generate a valid config even with serverless-appsync-plugin >= 2.2.0 3`] = `Array []`;

exports[`getAppSyncConfig should generate a valid config even with serverless-appsync-plugin >= 2.2.0 4`] = `
Array [
  Object {
    "invoke": [Function],
    "name": "lambda",
    "type": "AWS_LAMBDA",
  },
  Object {
    "config": Object {
      "accessKeyId": "DEFAULT_ACCESS_KEY",
      "endpoint": "http://localhost:8000",
      "region": "localhost",
      "secretAccessKey": "DEFAULT_SECRET",
      "sessionToken": "DEFAULT_SESSION_TOKEN",
      "tableName": "myTable",
    },
    "name": "dynamodb",
    "type": "AMAZON_DYNAMODB",
  },
  Object {
    "endpoint": "http://127.0.0.1",
    "name": "http",
    "type": "HTTP",
  },
]
`;

exports[`getAppSyncConfig should generate a valid config even with serverless-appsync-plugin >= 2.2.0 5`] = `Array []`;
